Add parsing from xContent to SearchProfileShardResults and nested classes

cbuescher opened the pull request:
In preparation for being able to parse SearchResponse from its rest representation
for the java rest client, this adds fromXContent to SearchProfileShardResults and its
nested classes.

cbuescher committed:
Adding parsing to ProfileResult

cbuescher committed:
Adding parsing to AggregationProfileShardResult

cbuescher committed:
Add parsing to QueryProfileShardResult

cbuescher committed:
Add parsing SearchProfileShardResults

cbuescher committed:
Small test changes concerning parsing time strings in ProfileResult and CollectorResult

cbuescher committed:
Adapt to introduction of the time_in_nanos field

javanna commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java:
```diff
@@ -376,12 +376,17 @@ public static void writeRawField(String field, BytesReference source, XContentBu
      * {@link XContentType}. Wraps the output into a new anonymous object.
      */
-    public static BytesReference toXContent(ToXContent toXContent, XContentType xContentType) throws IOException {
+    public static BytesReference toXContent(ToXContent toXContent, XContentType xContentType, boolean humanReadable) throws IOException {
```
nice, this method could be potentially called in all our tests so that we randomize the human readable flag, just to make sure we don't miss any of those additional fields. I think we haven't and this is the first time we encounter this but tests are the only way to actually verify. Maybe we can do this as a followup?

cbuescher replied:
++ good point

javanna replied:
can you look into this when you have time please?

javanna reviewed (changes requested):
left a few minors, thanks for doing this

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/ProfileResult.java:
```diff
@@ -162,4 +166,71 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        Map<String, Long> timings = new HashMap<>(parsedTimings.size());
+        for (Entry<String, Object> entry : parsedTimings.entrySet()) {
+            Object value = entry.getValue();
+            if (value instanceof Integer) {
```
And use `java.lang.Number`?

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/ProfileResult.java:
```diff
@@ -162,4 +166,71 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                }
+            }
+        }
+        Map<String, Long> timings = new HashMap<>(parsedTimings.size());
```
Can we check that `parsedTimings` is not null before doing this?

cbuescher replied:
This part has been removed while addressing another comment.

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/ProfileResult.java:
```diff
@@ -162,4 +166,71 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+     *            A map of breakdown timing for the node
+     * @return The total time at this node, inclusive of children
+     */
+    private static long getNodeTime(Map<String, Long> timings) {
```
What if `timings` is null? It doesn't seem to be required in the ctor

cbuescher replied:
Good point, I moved this code from somewhere where it was clear that the map is always there. I will add a check to the ctor.

cbuescher replied:
The reason I left the name was to make it clearer where the code was before in this PR (its from AbstractInternalProfileTree). I will rename it to `getTotalTime`. Also I think the comment is not up to date anymore (it doesn't sum any children) so I will remove that part)

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/SearchProfileShardResults.java:
```diff
@@ -95,6 +103,51 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser::getTokenLocation);
+        List<QueryProfileShardResult> queryProfileResults = new ArrayList<>();
+        AggregationProfileShardResult aggProfileShardResult = null;
+        String name = null;
```
Could this be renamed to `id` to avoid confusion?

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/SearchProfileShardResults.java:
```diff
@@ -95,6 +103,51 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            }
```
Maybe add another else block with unexpected token type?

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/query/CollectorResult.java:
```diff
@@ -153,4 +158,40 @@ public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params par
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            }
```
Same here, shouldn't we throw an exception if the token type is not expected? Or maybe I'm missing something.

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/query/QueryProfileShardResult.java:
```diff
@@ -90,15 +98,50 @@ public CollectorResult getCollectorResult() {
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            }
```
Maybe also here

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/SearchProfileShardResultsTests.java:
```diff
@@ -0,0 +1,79 @@
+        XContentType xcontentType = randomFrom(XContentType.values());
+        XContentBuilder builder = XContentFactory.contentBuilder(xcontentType);
+        builder.startObject();
+        builder = shardResult.toXContent(builder, ToXContent.EMPTY_PARAMS);
```
Can't we use XContentHelper.toXContent() here?

tlrx reviewed (changes requested):
Thanks @cbuescher, I'm sure that it hasn't been fun to do. I left some comments here and there, nothing big.

cbuescher committed:
Addressing review comments

cbuescher commented:
@tlrx @javanna thanks for the review, I pushed a commit that hopefully addresses all your comments.

javanna commented on core/src/main/java/org/elasticsearch/search/profile/SearchProfileShardResults.java:
```diff
@@ -98,15 +99,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         XContentParser.Token token = parser.currentToken();
         ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser::getTokenLocation);
-        Map<String, ProfileShardResult> searchProfileResults = new HashMap<>();
+        Map<String, ProfileShardResult> searchProfileResults = new LinkedHashMap<>();
```
why do we need ordering here?

cbuescher replied:
This way we preseve the order of the shard keys we encountered during xContent parsing. Without preserving this order, I encountered different xContent rendering outputs for the "shards" section (which is an array) in the roundtrip tests, so the order matters (at least the way `assertToXContentEquivalent` works so far).

cbuescher replied:
I pushed another commit that fixes the order in which the elements in the "shards" array are written out.

cbuescher committed:
Fixed order of ids in response shard section

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/query/QueryProfileShardResult.java:
```diff
@@ -90,15 +99,52 @@ public CollectorResult getCollectorResult() {
+        String currentFieldName = null;
+        List<ProfileResult> queryProfileResults = new ArrayList<>();
+        long rewriteTime = 0;
+        CollectorResult profileCollector = null;
```
Maybe profileCollector -> collector to reflect the field's name?

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/ProfileResultTests.java:
```diff
@@ -31,38 +34,86 @@
+        try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+            ProfileResult parsed = ProfileResult.fromXContent(parser);
+            assertEquals(profileResult.getTime(), parsed.getTime());
```
Can we check the parser in the try/catch block, and the parsed `ProfileResult` after that?

It's basically moving those two lines:
ProfileResult parsed = ProfileResult.fromXContent(parser);
assertEquals(profileResult.getTime(), parsed.getTime());

after the try block.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/ProfileResultTests.java:
```diff
@@ -31,38 +34,86 @@
-        timings.put("key1", 12345L);
-        timings.put("key2", 6789L);
-        ProfileResult result = new ProfileResult("someType", "some description", timings, children, 123456L);
+        children.add(new ProfileResult("child1", "desc1", timings1, Collections.emptyList()));
```
I suppose that you can use `singletonMap("key1", 100L)` directly here

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/SearchProfileShardResultsTests.java:
```diff
@@ -0,0 +1,74 @@
+public class SearchProfileShardResultsTests  extends ESTestCase {
+
+    public static SearchProfileShardResults createTestItem() {
+        int size = rarely() ? 0 : randomIntBetween(1, 2);
```
int size = randomIntBetween(0, 2) instead?

cbuescher replied:
Thats intentional, to get a small positiv value most of the time but occasionally test with 0. Otherwise it would be 1/3 which I think is too much.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/SearchProfileShardResultsTests.java:
```diff
@@ -0,0 +1,74 @@
+        Map<String, ProfileShardResult> searchProfileResults = new HashMap<>(size);
+        for (int i = 0; i < size; i++) {
+            List<QueryProfileShardResult> queryProfileResults = new ArrayList<>();
+            int queryItems = rarely() ? 0 : randomIntBetween(0, 2);
```
`int queryItems = randomIntBetween(0, 2);` is simplier?

cbuescher replied:
Same here, I'd like to keep that.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/aggregation/AggregationProfileShardResultTests.java:
```diff
@@ -0,0 +1,93 @@
+    public void testFromXContent() throws IOException {
+        AggregationProfileShardResult profileResult = createTestItem(2);
+        XContentType xcontentType = randomFrom(XContentType.values());
+        XContentBuilder builder = XContentFactory.contentBuilder(xcontentType);
```
Can we use the `XContentBuilder` inside a try-with-resource block?

This is usually how they are used and it has the effect to close the builder and fail fast if the xcontent object is invalid (see JsonXContentGenerator.close() method)

cbuescher replied:
I will switch to using the `toXContent()` helper here like in other tests.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/aggregation/AggregationProfileShardResultTests.java:
```diff
@@ -0,0 +1,93 @@
+        XContentParserUtils.ensureFieldName(parser, parser.nextToken(), AggregationProfileShardResult.AGGREGATIONS);
+        XContentParserUtils.ensureExpectedToken(parser.nextToken(), XContentParser.Token.START_ARRAY, parser::getTokenLocation);
+        AggregationProfileShardResult parsed = AggregationProfileShardResult.fromXContent(parser);
+        assertToXContentEquivalent(builder.bytes(), toXContent(parsed, xcontentType), xcontentType);
```
Can we move this down, after the `Token` checks?

cbuescher replied:
done

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/aggregation/AggregationProfileShardResultTests.java:
```diff
@@ -0,0 +1,93 @@
+        ProfileResult profileResult = new ProfileResult("someType", "someDescription", timings, Collections.emptyList());
+        profileResults.add(profileResult);
+        AggregationProfileShardResult aggProfileResults = new AggregationProfileShardResult(profileResults);
+        XContentBuilder builder = JsonXContent.contentBuilder();
```
Same here, try with resource block

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/query/CollectorResultTests.java:
```diff
@@ -22,15 +22,55 @@
+        CollectorResult collectorResult = createTestItem(1);
+        XContentType xcontentType = randomFrom(XContentType.values());
+        XContentBuilder builder = XContentFactory.contentBuilder(xcontentType);
+        boolean humanReadable = randomBoolean();
```
No need to use an intermediate variable for this

cbuescher replied:
I need it later when comparing to the xContent rendering of the parsed object.

tlrx replied:
I meant `builder.humanReadable(randomBoolean());` here and use `builder.humanReadable() ` later

cbuescher replied:
Ah, that doesn't work after my last changes because I got rid of the builder.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/query/QueryProfileShardResultTests.java:
```diff
@@ -0,0 +1,66 @@
+            queryProfileResults.add(ProfileResultTests.createTestItem(1));
+        }
+        CollectorResult profileCollector = CollectorResultTests.createTestItem(2);
+        long rewriteTime = randomNonNegativeLong();
```
Should we use the `%100_000` strategy here too?

cbuescher replied:
Yes, good point.

tlrx commented on core/src/test/java/org/elasticsearch/search/profile/query/QueryProfileShardResultTests.java:
```diff
@@ -0,0 +1,66 @@
+    public void testFromXContent() throws IOException {
+        QueryProfileShardResult profileResult = createTestItem();
+        XContentType xcontentType = randomFrom(XContentType.values());
+        XContentBuilder builder = XContentFactory.contentBuilder(xcontentType);
```
Same here

tlrx reviewed (changes requested):
Thanks for the hard work! I made many comments but they are all cosmetics or minor. This is getting very close

tlrx commented on core/src/main/java/org/elasticsearch/search/profile/SearchProfileShardResults.java:
```diff
@@ -75,26 +87,71 @@ public void writeTo(StreamOutput out) throws IOException {
-
-        for (Map.Entry<String, ProfileShardResult> entry : shardResults.entrySet()) {
+        builder.startObject(PROFILE_FIELD).startArray(SHARDS_FIELD);
+        TreeSet<String> sortedKeys = new TreeSet<>(shardResults.keySet());
```
Are we forced to sort all shard results every time just to be able to check them in tests? Can't we check / sort again in the test instead?

cbuescher replied:
I talked to @javanna about this, the idea wasto also make the rendered response reproducibly the same for the same source object.

javanna replied:
it is not just for tests. Unfortunately this object uses a map to store information that goes into a json array, which is by definition ordered. Either we change the format of the response or we guarantee ordering then. I think this is a good way to guarantee ordering and make the response more human readable too.

javanna replied:
maybe add a comment to explain this @cbuescher ?

cbuescher commented on core/src/main/java/org/elasticsearch/search/profile/ProfileResult.java:
```diff
@@ -162,4 +166,68 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                } else if (DESCRIPTION.match(currentFieldName)) {
+                    description = parser.text();
+                } else if (NODE_TIME.match(currentFieldName)) {
+                    // we need to consume this value, but we use the raw nanosecond value
```
Yes, we sum the timings values in ProfileResult ctor now, no need for both of these values, I'll update the comments

cbuescher committed:
Addressing more review comments

cbuescher commented:
@tlrx thanks for reviewing, I added a commit that addresses most of your comments and left some answer where I'd like to keep things (or discuss further)

javanna reviewed (approved):
LGTM

cbuescher committed:
adding comments

tlrx reviewed (approved):
LGTM

cbuescher committed:
minor addition

cbuescher commented:
Ported to 5.x with c8e027fc9857cf8154f7e7d5ac2622585b5b3083
